### PR TITLE
feat: add endpoint struct to ServiceConfigEntry

### PR DESCRIPTION
**Pull request by DanStough**
### Description
First PR of many covering new terminating gateway functionality allowing external traffic w/o node registration and transparent proxying enabled.

This PR adds an `endpoint` field to `service-defaults` which represents a service external to Consul. It can be a hostname (w/ or w/o a wildcard), ipv4/ipv6 address or ipv4/ipv6 CIDR. Service defaults will represent the "virtual service", which will be consumed by the existing `Services` definition in `TerminatingGatewayConfigEntry`.

### Testing & Reproduction steps
Unit tests added. Higher-level testing will happen when we start integrating the watch with the terminating gateway xDS generation.

### PR Checklist

* [X] updated test coverage
* [ ] ~external facing docs updated~
* [X] not a security concern
* [X] checklist [folder](./../docs/config) consulted

### CODE REVIEW

File: agent/structs/config_entry.go [hunk 1]
```diff
@@ -253,6 +320,30 @@ func (c *UpstreamConfiguration) Clone() *UpstreamConfiguration {
 	return &c2
 }
 
+// EndpointConfig represents a virtual service, i.e. one that is external to Consul
+type EndpointConfig struct {
```
Comment (DanStough) [inline agent/structs/config_entry.go:324]: I left out protocol since it's already part of `service-defaults` and I think it's confusing to have it in two places.

I left in the TLS stuff because I think the spirit of this is that the tls authorization should be at the service level and not the gateway level.

File: agent/structs/config_entry.go [hunk 2]
```diff
@@ -190,9 +199,67 @@ func (e *ServiceConfigEntry) Validate() error {
 		}
 	}
 
+	if e.Endpoint != nil {
+		gatewayTLSEnabled := e.Endpoint.CAFile != ""
+
+		// TODO (dans): need to find a way to validate this against the proxy defaults.
+		if gatewayTLSEnabled && e.Mode == ProxyModeTransparent {
+			validationErr = multierror.Append(validationErr, errors.New("Certificates can not be defined in an endpoint when proxy is in transparent mode; these are the responsibility of the service dialer"))
```
Comment (DanStough) [inline agent/structs/config_entry.go:207]: Based on what we talked about earlier in the week.

**Comment by DanStough** [comment]
I can't really explain why, but bumping the version of `bats` solved the `integration-test` [problems](https://app.circleci.com/pipelines/github/hashicorp/consul/31652/workflows/f550327a-535d-4b3f-a910-2bea75a1920a/jobs/687002) 🤷‍♂️ .

File: agent/structs/config_entry_test.go [hunk 3]
```diff
@@ -427,6 +427,48 @@ func TestDecodeConfigEntry(t *testing.T) {
 				},
 			},
 		},
+		{
```
Comment (rboyer) [inline agent/structs/config_entry_test.go:430]: Kitchen sink config entry tests exist in 3 places. You're only missing `command/config/write/config_write_test.go`

File: test/integration/connect/envoy/Dockerfile-bats [hunk 4]
```diff
@@ -1,6 +1,6 @@
 FROM docker.mirror.hashicorp.services/fortio/fortio AS fortio
 
-FROM docker.mirror.hashicorp.services/bats/bats:1.6.0
```
Comment (rboyer) [inline test/integration/connect/envoy/Dockerfile-bats:3]: Why did you need to bump this?

File: agent/structs/config_entry.go (hunk 1 above)
Comment (rboyer) [reply agent/structs/config_entry.go:318]: I think the TLS stuff should probably still continue to live in the gateway config entry.

These reference files that must already exist on the machine. `service-defaults` sorta represents the service in the abstract and may be used in a variety of consuming services on many machines whereas the terminating gateway config entry largely exists 1:1 with some actual real instance of the gateway running on a machine or small set of machines where you can arrange for those files to exist.

File: test/integration/connect/envoy/Dockerfile-bats (hunk 4 above)
Comment (DanStough) [reply test/integration/connect/envoy/Dockerfile-bats:3]: See above https://github.com/hashicorp/consul/pull/13159#issuecomment-1133395868

**Review by rboyer — APPROVED** [review]
LGTM with non blocking comment about a dead function argument.

File: agent/structs/config_entry.go [hunk 5]
```diff
@@ -190,9 +199,38 @@ func (e *ServiceConfigEntry) Validate() error {
...
+
+		if e.Endpoint.Port < 1 || e.Endpoint.Port > 65535 {
+			validationErr = multierror.Append(validationErr, fmt.Errorf("Invalid Port number %d", e.Endpoint.Port))
+		}
+	}
+
 	return validationErr
 }
 
+func validateEndpointAddress(tlsEnabled bool, address string) error {
```
[the remark this answers is not in the document]
Comment (DanStough) [reply agent/structs/config_entry.go:216]: Nope, good call. Will remove

**Review by rboyer — APPROVED** [review]
